# **Segmenting and Clustering Neighborhoods in Toronto**

### Get the packages we need for to start

In [282]:
#!pip install beautifulsoup4
#!pip install lxml
#!pip install html5lib
#!pip install requests

import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup as bsp

Fetching the datafile with *requests.get()* and create the soup using *'lxml'* as parser
Then find the table object in the soup and create pandas dataframe from it
The original dataset has a column named **'PostCode'**, but according to the request we need to change it to **'PostalCode'**

In [283]:
source = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")
soup = bsp(source.content, 'lxml')
table = soup.find_all('table')[0] 
df = pd.read_html(str(table))
#type(df)
#df[0]

df2=pd.DataFrame(df[0])  
df2.columns = ['PostalCode','Borough','Neighborhood']
df2.head()

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

In [284]:
df2.shape

(289, 3)

Looks good, but we need to find all the rows where the value of **'Borough1** are equal to **'Not assigned'** and drop them

In [285]:
df2[df2['Borough'] == 'Not assigned'].head()

PostalCode       Borough  Neighborhood
0         M1A  Not assigned  Not assigned
1         M2A  Not assigned  Not assigned
9         M8A  Not assigned  Not assigned
13        M2B  Not assigned  Not assigned
20        M7B  Not assigned  Not assigned

I'm going to find all the rows where 'Borough' is **not equal** to 'Not assigned' and with that rows I overwrite the existing dataframe

In [286]:
df2 = df2[df2['Borough'] != 'Not assigned']
df2.head(10)

PostalCode           Borough      Neighborhood
2         M3A        North York         Parkwoods
3         M4A        North York  Victoria Village
4         M5A  Downtown Toronto      Harbourfront
5         M5A  Downtown Toronto       Regent Park
6         M6A        North York  Lawrence Heights
7         M6A        North York    Lawrence Manor
8         M7A      Queen's Park      Not assigned
10        M9A         Etobicoke  Islington Avenue
11        M1B       Scarborough             Rouge
12        M1B       Scarborough           Malvern

For the sake of certainty we should check whether the dataframe does not contain any rows like that

In [287]:
df2[df2['Borough'] == 'Not assigned']

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

In [288]:
df2.shape

(212, 3)

so we started with 289 rows and now we have 212.

The next thing we should handle is to find all the rows where **'Neighborhood'** are equal to **'Not assigned'**

In [289]:
df2[df2['Neighborhood']=='Not assigned']

PostalCode       Borough  Neighborhood
8        M7A  Queen's Park  Not assigned

In [290]:
df2.loc[df2['Neighborhood'] == 'Not assigned', ['Borough']].values

array([["Queen's Park"]], dtype=object)

first we fetch all the involved rows then assign to its 'Neighborhood' column the *value* of the the 'Borough' column

In [291]:
df2.loc[df2['Neighborhood'] == 'Not assigned', ['Neighborhood']] = df2.loc[df2['Neighborhood'] == 'Not assigned', ['Borough']].values

Then check the results whether the value of the 'Neighborhood' column has changed in 8th row

In [292]:
df2.head(10)

PostalCode           Borough      Neighborhood
2         M3A        North York         Parkwoods
3         M4A        North York  Victoria Village
4         M5A  Downtown Toronto      Harbourfront
5         M5A  Downtown Toronto       Regent Park
6         M6A        North York  Lawrence Heights
7         M6A        North York    Lawrence Manor
8         M7A      Queen's Park      Queen's Park
10        M9A         Etobicoke  Islington Avenue
11        M1B       Scarborough             Rouge
12        M1B       Scarborough           Malvern

In [293]:
df2.shape

(212, 3)

Now we have to combine the rows where 'Neighborhoods' go to the same 'PostalCode' 

For that I used the *groupby()* function along with the *.agg()* method

The result is *almost* the desired table, but we have to handle the 'Borough' column witch might consist the name of the 'Borough' more than once now

So I'm using *.split()* method to split them and put the results in two column, so we have the firt occurance in the first column and all of the others in the second column.

I drop the second column because we need the values only once

I changed the values of the 'Borough' column to the results in the dataframe

Check the results and see whether it is okay or not

In [294]:
from pandas import Series

df3=df2.groupby('PostalCode').agg({'Borough':lambda x:', '.join(x)
                        ,'Neighborhood':lambda x:', '.join(x)
                        })
df3.reset_index(inplace=True)
s=df3['Borough'].str.split(', ',1).apply(Series,1)
s.drop(columns=1,inplace=True)

df3['Borough']= s

df3[df3['PostalCode']=='M5A']



PostalCode           Borough               Neighborhood
53        M5A  Downtown Toronto  Harbourfront, Regent Park

Check the dataframe:

In [295]:
df3.shape

(103, 3)

## End of part one
------


-----

## Part two

To get the geographical coordinates from the file I'm going to use pandas *.read_csv()* method

In [296]:
geo = pd.read_csv("http://cocl.us/Geospatial_data")
geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Let's merge the to dataframes

In [297]:
df4 = pd.concat([df3,geo],axis=1, sort=False)
df4.head()

PostalCode      Borough                            Neighborhood Postal Code  \
0        M1B  Scarborough                          Rouge, Malvern         M1B   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union         M1C   
2        M1E  Scarborough       Guildwood, Morningside, West Hill         M1E   
3        M1G  Scarborough                                  Woburn         M1G   
4        M1H  Scarborough                               Cedarbrae         M1H   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

Drop the second 'Postal Code' column

In [298]:
df4= df4[['PostalCode','Borough','Neighborhood','Latitude','Longitude']]
df4.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

## End of part two
------


-----

## Part three

Explore and cluster the neighborhoods in Toronto

Also I decided to work with only boroughs that contain the word Toronto

In [299]:
df4['Borough'].value_counts()

North York          24
Downtown Toronto    18
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East Toronto         5
East York            5
York                 5
Mississauga          1
Queen's Park         1
Name: Borough, dtype: int64

#### Use geopy library to get the latitude and longitude values of Toronto.

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>TO_explorer</em>, as shown below.

In [302]:
from geopy.geocoders import Nominatim
import folium

address = 'Toronto, ON'

geolocator = Nominatim(user_agent="TO_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


Create map of Toronto using the given latitude and longitude values

In [303]:

map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df4['Latitude'], df4['Longitude'], df4['Borough'], df4['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Simplify the above map and segment and cluster only boroughs that contain the word Toronto. So let's slice the original dataframe and create a new dataframe of the Toronto data.

In [304]:
Toronto_data = df4[df4['Borough'].str.contains('Toronto')]
Toronto_data.reset_index(drop=True, inplace=True)
Toronto_data.head()

PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [305]:
df4[df4['Borough'].str.contains('Toronto')]['Borough'].unique()

array(['East Toronto', 'Central Toronto', 'Downtown Toronto',
       'West Toronto'], dtype=object)

Get the geographical coordinates of the involved Boroughs

In [306]:
address_list=[]
for i in df4[df4['Borough'].str.contains('Toronto')]['Borough'].unique():
    address_list.append(i + ', ON')
print(address_list)

['East Toronto, ON', 'Central Toronto, ON', 'Downtown Toronto, ON', 'West Toronto, ON']


Visualize the neighborhoods, but we should adjust the previously used code because we have mor than on address this time

In [307]:
address = address_list

for i in address:
    geolocator = Nominatim(user_agent="TO_explorer")
    location = geolocator.geocode(i)
    latitude = location.latitude
    longitude = location.longitude
    #print('The geograpical coordinate of {} are {}, {}.'.format(i, latitude, longitude))
    map_downtownT = folium.Map(location=[latitude, longitude], zoom_start=11)  
    for lat, lng, label in zip(downtown_data['Latitude'], downtown_data['Longitude'], downtown_data['Neighborhood']):
        label = folium.Popup(label, parse_html=True)
        folium.CircleMarker(
            [lat, lng],
            radius=5,
            popup=label,
            color='blue',
            fill=True,
            fill_color='#3186cc',
            fill_opacity=0.7,
            parse_html=False).add_to(map_downtownT)  
map_downtownT
    
    

-----

#### Define Foursquare Credentials and Version

I'm going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

In [337]:
CLIENT_ID = '************************************' # your Foursquare ID
CLIENT_SECRET = '**************************************' # your Foursquare Secret
VERSION = '20190224' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: ************************************
CLIENT_SECRET:**************************************


-------

#### Explore the first neighborhood

Get the neighborhood's name, latitude and longitude values.

In [309]:
Toronto_data.loc[0, 'Neighborhood']

'The Beaches'

In [310]:
neighborhood_latitude = Toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = Toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = Toronto_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312.


#### Now, let's get the top 100 venues that are in 'The Beaches' within a radius of 500 meters.

First, let's create the GET request URL. 

In [311]:
radius=500
LIMIT=100
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    VERSION, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=1HN2DYV3VM2MLMKC2DRYIW14O22AMS0UMIL4EZJY0MTQZOJB&client_secret=WSDV1HNYYU3NUMX1FNDGN3I0LA5JPP541NZQPPV5IW4MC03L&ll=43.67635739999999,-79.2930312&v=20190224&radius=500&limit=100'

Send the GET request and examine the resutls

In [312]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5c72c1694c1f6762ddc47711'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'The Beaches',
  'headerFullLocation': 'The Beaches, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 4,
  'suggestedBounds': {'ne': {'lat': 43.680857404499996,
    'lng': -79.28682091449052},
   'sw': {'lat': 43.67185739549999, 'lng': -79.29924148550948}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4ad4c062f964a52011f820e3',
       'name': 'The Big Carrot Natural Food Market',
       'location': {'address': '125 Southwood Dr',
        'lat': 43.678879,
        'lng': -79.297734,
        'labeledLatLngs': [{'label': 'display',
          'lat

In [313]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now clean the json and structure it into a *pandas* dataframe.

In [314]:
import json
from pandas.io.json import json_normalize


venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name         categories        lat        lng
0  The Big Carrot Natural Food Market  Health Food Store  43.678879 -79.297734
1                 Grover Pub and Grub                Pub  43.679181 -79.297215
2                           Starbucks        Coffee Shop  43.678798 -79.298045
3                       Upper Beaches       Neighborhood  43.680563 -79.292869

In [315]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

4 venues were returned by Foursquare.


-------
## 2. Explore involved Boroughs in Toronto

#### Let's create a function to repeat the same process to the involved Boroughs in Toronto

In [316]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Run the above function on each involved Borough and create a new dataframe called *toronto_venues*.

In [318]:

toronto_venues = getNearbyVenues(names=Toronto_data['Neighborhood'],
                                   latitudes=Toronto_data['Latitude'],
                                   longitudes=Toronto_data['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North, Forest Hill West
The Annex, North Midtown, Yorkville
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, Exhibition Place, Parkdale Village
High Park, The 

In [319]:
print(toronto_venues.shape)
toronto_venues.head()

(1699, 7)


Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031  The Big Carrot Natural Food Market       43.678879   
1              -79.293031                 Grover Pub and Grub       43.679181   
2              -79.293031                           Starbucks       43.678798   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                            Pantheon       43.677621   

   Venue Longitude     Venue Category  
0       -79.297734  Health Food Store  
1       -79.297215                Pub  
2       -79.298045        Coffee Shop  
3       -79.292869       Neighborhood  
4       -79.351434   Greek Restaurant

Let's check how many venues were returned for each neighborhood

In [320]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            57   
Brockton, Exhibition Place, Parkdale Village                           19   
Business Reply Mail Processing Centre 969 Eastern                      18   
CN Tower, Bathurst Quay, Island airport, Harbou...                     13   
Cabbagetown, St. James Town                                            43   
Central Bay Street                                                     81   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               16   
Church and Wellesley                                                   86   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             38   
Davisville North                                                        9   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     15   
Design Exchange, Toronto Dominion Centre                              100   
Dovercourt Village, Dufferin                                           20   
First Canadian Place, Underground city                                100   
Forest Hill North, Forest Hill West                                     4   
Harbord, University of Toronto                                         35   
Harbourfront East, Toronto Islands, Union Station                     100   
Harbourfront, Regent Park                                              47   
High Park, The Junction South                                          23   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               63   
Moore Park, Summerhill East                                             3   
North Toronto West                                                     21   
Parkdale, Roncesvalles                                                 16   
Rosedale                                                                4   
Roselawn                                                                3   
Runnymede, Swansea                                                     38   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        94   
Studio District                                                        40   
The Annex, North Midtown, Yorkville                                    24   
The Beaches                                                             4   
The Beaches West, India Bazaar                                         19   
The Danforth West, Riverdale                                           42   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             57   
Brockton, Exhibition Place, Parkdale Village                            19   
Business Reply Mail Processing Centre 969 Eastern                       18   
CN Tower, Bathurst Quay, Island airport, Harbou...                      13   
Cabbagetown, St. James Town                                             43   
Central Bay Street                                                      81   
Chinatown, Grange Park, Kensington Market                              100   
Christie                                                                16   
Church and Wellesley                                                    86   
Commerce Court, Victoria Hotel     

#### Unique categories 

In [321]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 237 uniques categories.


------

## 3. Analyze Each Neighborhood

In [322]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
# I need to drop it, becasue it is placed in alphabetical order but I want it in the first column
toronto_onehot.drop('Neighborhood', axis=1, inplace=True)
# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 
# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()


Neighborhood  Adult Boutique  Afghan Restaurant  Airport  \
0                   The Beaches               0                  0        0   
1                   The Beaches               0                  0        0   
2                   The Beaches               0                  0        0   
3                   The Beaches               0                  0        0   
4  The Danforth West, Riverdale               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Trail  Train Station  \
0                 0                    0  ...      0              0   
1                 0                    0  ...      0              0   
2                 0                    0  ...      0              0   
3                 0                    0  ...      0              0   
4                 0                    0  ...      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0         0          0            0              0            0  
1         0          0            0              0            0  
2         0          0            0              0            0  
3         0          0            0              0            0  
4         0          0            0              0            0  

[5 rows x 237 columns]

In [323]:
toronto_onehot.shape

(1699, 237)

#### Group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [324]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Adult Boutique  \
0                            Adelaide, King, Richmond        0.000000   
1                                         Berczy Park        0.000000   
2        Brockton, Exhibition Place, Parkdale Village        0.000000   
3   Business Reply Mail Processing Centre 969 Eastern        0.000000   
4   CN Tower, Bathurst Quay, Island airport, Harbo...        0.000000   
5                         Cabbagetown, St. James Town        0.000000   
6                                  Central Bay Street        0.000000   
7           Chinatown, Grange Park, Kensington Market        0.000000   
8                                            Christie        0.000000   
9                                Church and Wellesley        0.011628   
10                     Commerce Court, Victoria Hotel        0.000000   
11                                         Davisville        0.000000   
12                                   Davisville North        0.000000   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...        0.000000   
14           Design Exchange, Toronto Dominion Centre        0.000000   
15                       Dovercourt Village, Dufferin        0.000000   
16             First Canadian Place, Underground city        0.000000   
17                Forest Hill North, Forest Hill West        0.000000   
18                     Harbord, University of Toronto        0.000000   
19  Harbourfront East, Toronto Islands, Union Station        0.000000   
20                          Harbourfront, Regent Park        0.000000   
21                      High Park, The Junction South        0.000000   
22                                      Lawrence Park        0.000000   
23                           Little Portugal, Trinity        0.000000   
24                        Moore Park, Summerhill East        0.000000   
25                                 North Toronto West        0.000000   
26                             Parkdale, Roncesvalles        0.000000   
27                                           Rosedale        0.000000   
28                                           Roselawn        0.000000   
29                                 Runnymede, Swansea        0.000000   
30                           Ryerson, Garden District        0.000000   
31                                     St. James Town        0.000000   
32                    Stn A PO Boxes 25 The Esplanade        0.000000   
33                                    Studio District        0.000000   
34                The Annex, North Midtown, Yorkville        0.000000   
35                                        The Beaches        0.000000   
36                     The Beaches West, India Bazaar        0.000000   
37                       The Danforth West, Riverdale        0.000000   

    Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0            0.000000  0.000000            0.000000      0.000000   
1            0.000000  0.000000            0.000000      0.000000   
2            0.000000  0.000000            0.000000      0.000000   
3            0.000000  0.000000            0.000000      0.000000   
4            0.000000  0.076923            0.076923      0.076923   
5            0.000000  0.000000            0.000000      0.000000   
6            0.000000  0.000000            0.000000      0.000000   
7            0.000000  0.000000            0.000000      0.000000   
8            0.000000  0.000000            0.000000      0.000000   
9            0.011628  0.000000            0.000000      0.000000   
10           0.000000  0.000000            0.000000      0.000000   
11           0.000000  0.000000            0.000000      0.000000   
12           0.000000  0.000000            0.000000      0.000000   
13           0.000000  0.000000            0.000000      0.000000   
14           0.000000  0.000000            0.000000      0.000000   
15           0.000000  0.000000            0.000000      0.000000   
16           0.00000

In [325]:
toronto_grouped.shape

(38, 237)

#### Let's print each neighborhood along with the top 5 most common venues

In [326]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
                 venue  freq
0          Coffee Shop  0.06
1                 Café  0.05
2           Steakhouse  0.04
3  American Restaurant  0.04
4      Thai Restaurant  0.04


----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2          Restaurant  0.05
3  Italian Restaurant  0.04
4                Café  0.04


----Brockton, Exhibition Place, Parkdale Village----
                  venue  freq
0        Breakfast Spot  0.11
1                  Café  0.11
2           Coffee Shop  0.11
3  Gym / Fitness Center  0.05
4  Caribbean Restaurant  0.05


----Business Reply Mail Processing Centre 969 Eastern----
              venue  freq
0       Yoga Studio  0.06
1     Auto Workshop  0.06
2              Park  0.06
3        Comic Shop  0.06
4  Recording Studio  0.06


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
              venue  freq
0    Airp

Write a function to sort the venues in descending order.

In [327]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Create the new dataframe and display the top 10 venues for each neighborhood.

In [328]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village                  Café   
3  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
4  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue   4th Most Common Venue  \
0                  Café   American Restaurant              Steakhouse   
1          Cocktail Bar            Restaurant                  Bakery   
2        Breakfast Spot           Coffee Shop  Furniture / Home Store   
3      Recording Studio            Smoke Shop              Skate Park   
4       Airport Service      Airport Terminal        Sculpture Garden   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Thai Restaurant        Clothing Store                 Hotel   
1             Steakhouse    Italian Restaurant        Farmers Market   
2  Performing Arts Venue    Italian Restaurant          Climbing Gym   
3                Brewery         Burrito Place               Butcher   
4        Harbor / Marina               Airport    Airport Food Court   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant                   Bar                 Bakery  
1              Beer Bar           Cheese Shop                    Pub  
2               Stadium  Caribbean Restaurant          Burrito Place  
3            Restaurant            Comic Shop          Auto Workshop  
4          Airport Gate              Boutique          Boat or Ferry

-----

## 4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [329]:
from sklearn.cluster import KMeans
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([3, 3, 3, 3, 3, 3, 3, 3, 3, 3])

Create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [330]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = Toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Toronto_merged.head() # check the last columns!

PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               4     Health Food Store           Coffee Shop   
1 -79.352188               3      Greek Restaurant           Coffee Shop   
2 -79.315572               3        Sandwich Place                   Gym   
3 -79.340923               3                  Café           Coffee Shop   
4 -79.388790               1                  Park           Swim School   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Pub                 Diner  Fast Food Restaurant   
1        Ice Cream Shop             Bookstore    Italian Restaurant   
2    Italian Restaurant             Pet Store           Pizza Place   
3    Italian Restaurant                Bakery   American Restaurant   
4    Dim Sum Restaurant              Bus Line           Yoga Studio   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0        Farmers Market    Falafel Restaurant           Event Space   
1        Cosmetics Shop               Brewery       Bubble Tea Shop   
2                   Pub         Movie Theater         Burrito Place   
3           Yoga Studio                  Park    Seafood Restaurant   
4               Dog Run  Fast Food Restaurant        Farmers Market   

  9th Most Common Venue 10th Most Common Venue  
0  Ethiopian Restaurant      Electronics Store  
1  Caribbean Restaurant             Restaurant  
2          Burger Joint                Brewery  
3        Sandwich Place            Cheese Shop  
4    Falafel Restaurant            Event Space

Visualize the resulting clusters

In [331]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighborhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

----
## 5. Examine Clusters

Now examine each cluster and determine the discriminating venue categories that distinguish each cluster. 

#### Cluster 1

In [332]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22  Central Toronto               0           Music Venue   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22          Home Service                Garden           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22  Fast Food Restaurant        Farmers Market    Falafel Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22           Event Space  Ethiopian Restaurant      Electronics Store

#### Cluster 2

In [333]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
4    Central Toronto               1                  Park   
10  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4            Swim School    Dim Sum Restaurant              Bus Line   
10            Playground                 Trail                 Diner   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4            Yoga Studio               Dog Run  Fast Food Restaurant   
10  Fast Food Restaurant        Farmers Market    Falafel Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
4         Farmers Market    Falafel Restaurant            Event Space  
10           Event Space  Ethiopian Restaurant      Electronics Store

#### Cluster 3

In [334]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
8  Central Toronto               2            Playground   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
8                   Gym            Restaurant                 Diner   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
8  Fast Food Restaurant        Farmers Market    Falafel Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
8           Event Space  Ethiopian Restaurant      Electronics Store

#### Cluster 4

In [335]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1       East Toronto               3      Greek Restaurant   
2       East Toronto               3        Sandwich Place   
3       East Toronto               3                  Café   
5    Central Toronto               3          Burger Joint   
6    Central Toronto               3        Clothing Store   
7    Central Toronto               3        Sandwich Place   
9    Central Toronto               3     Convenience Store   
11  Downtown Toronto               3           Coffee Shop   
12  Downtown Toronto               3   Japanese Restaurant   
13  Downtown Toronto               3           Coffee Shop   
14  Downtown Toronto               3           Coffee Shop   
15  Downtown Toronto               3           Coffee Shop   
16  Downtown Toronto               3           Coffee Shop   
17  Downtown Toronto               3           Coffee Shop   
18  Downtown Toronto               3           Coffee Shop   
19  Downtown Toronto               3           Coffee Shop   
20  Downtown Toronto               3           Coffee Shop   
21  Downtown Toronto               3           Coffee Shop   
23   Central Toronto               3    Mexican Restaurant   
24   Central Toronto               3           Coffee Shop   
25  Downtown Toronto               3                  Café   
26  Downtown Toronto               3                  Café   
27  Downtown Toronto               3        Airport Lounge   
28  Downtown Toronto               3           Coffee Shop   
29  Downtown Toronto               3                  Café   
30  Downtown Toronto               3         Grocery Store   
31      West Toronto               3           Supermarket   
32      West Toronto               3                   Bar   
33      West Toronto               3                  Café   
34      West Toronto               3                  Café   
35      West Toronto               3        Breakfast Spot   
36      West Toronto               3           Coffee Shop   
37      East Toronto               3           Yoga Studio   

   2nd Most Common Venue  3rd Most Common Venue  \
1            Coffee Shop         Ice Cream Shop   
2                    Gym     Italian Restaurant   
3            Coffee Shop     Italian Restaurant   
5                    Gym          Grocery Store   
6            Coffee Shop    Sporting Goods Shop   
7           Dessert Shop            Pizza Place   
9                    Pub            Coffee Shop   
11            Restaurant                 Bakery   
12           Coffee Shop       Sushi Restaurant   
13                Bakery                   Park   
14        Clothing Store                   Café   
15            Restaurant                  Hotel   
16          Cocktail Bar             Restaurant   
17                  Café     Italian Restaurant   
18                  Café    American Restaurant   
19                 Hotel               Aquarium   
20                  Café                  Hotel   
21                  Café             Restaurant   
23                 Trail       Sushi Restaurant   
24        Sandwich Place                   Café   
25           Coffee Shop                 Bakery   
26                   Bar  Vietnamese Restaurant   
27       Airport Service       Airport Terminal   
28            Restaurant                   Café   
29           Coffee Shop                  Hotel   
30                  Café                   Park   
31                Bakery               Pharmacy   
32           Men's Store       Asian Restaurant   
33        Breakfast Spot            Coffee Shop   
34    Mexican Restaurant            Flea Market   
35             Gift Shop                Dog Run   
36                  Café            Pizza Place   
37      Recording Studio             Smoke Shop   

            4th Most Common Venue      5th Most Common Venue  \
1                       Bookstore         Italian Restaurant   
2                       Pet Store        

#### Cluster 5

In [336]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  East Toronto               4     Health Food Store           Coffee Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Pub                 Diner  Fast Food Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0        Farmers Market    Falafel Restaurant           Event Space   

  9th Most Common Venue 10th Most Common Venue  
0  Ethiopian Restaurant      Electronics Store